app/org/sagebionetworks/bridge/upload/UploadUtil.java

```diff
@@ -1,17 +1,250 @@
+            }
+            case STRING: {
+                // If the value isn't a string, trivially convert it into a string.
+                return CanonicalizationResult.makeResult(convertToStringNode(valueNode));
+            }
+            case TIME_V2: {
+                if (!valueNode.isTextual()) {
+                    return CanonicalizationResult.makeError("Invalid time JSON value " + valueNode.toString());
+                }
+
+                // This is a time without date or time-zone, akin to Joda LocalTime. First parse it as a LocalTime.
+                String timeStr = valueNode.textValue();
+                LocalTime parsedLocalTime = null;
+                try {
+                    parsedLocalTime = LocalTime.parse(timeStr);
+                } catch (IllegalArgumentException ex) {
+                    // Swallow exception. We have better logging later in the chain.
+                }
+
+                if (parsedLocalTime == null) {
+                    // If that doesn't work, fall back to parsing a full timestamp and use just the LocalTime part.
+                    DateTime parsedDateTime = parseIosTimestamp(timeStr);
+                    if (parsedDateTime != null) {
+                        parsedLocalTime = parsedDateTime.toLocalTime();
+                    }
+                }
+
+                if (parsedLocalTime != null) {
+                    return CanonicalizationResult.makeResult(new TextNode(parsedLocalTime.toString()));
+                } else {
+                    return CanonicalizationResult.makeError("Invalid time string " + timeStr);
+                }
+            }
+            case TIMESTAMP: {
+                if (valueNode.isNumber()) {
+                    // If this is a number, then it's epoch milliseconds (implicitly in UTC).
+                    return CanonicalizationResult.makeResult(new TextNode(DateUtils.convertToISODateTime(
+                            valueNode.longValue())));
+                } else if (valueNode.isTextual()) {
+                    String dateTimeStr = valueNode.textValue();
+                    DateTime parsedDateTime = parseIosTimestamp(dateTimeStr);
+                    if (parsedDateTime != null) {
+                        return CanonicalizationResult.makeResult(new TextNode(parsedDateTime.toString()));
+                    } else {
+                        return CanonicalizationResult.makeError("Invalid date-time (timestamp) string " + dateTimeStr);
+                    }
+                } else {
+                    return CanonicalizationResult.makeError("Invalid date-time (timestamp) JSON value " +
+                            valueNode.toString());
+                }
+            }
+            default: {
+                // Should never happen, but just in case.
+                return CanonicalizationResult.makeError("Unknown field type " + type.name());
+            }
+        }
```

Comment:
If this continues to grow, perhaps a map between type and a handler interface with the implementation for each type.